cloud_utils/vault.py

```diff
@@ -1,136 +1,35 @@
-    token: Callable[[], Coroutine[Any, Any, str]],
-) -> Callable[[], Coroutine[Any, Any, Dict[str, str]]]:
-    async def _vault_headers() -> Dict[str, str]:
-        return {
-            "X-Vault-Token": await token(),
-        }
-
-    return _vault_headers
-
-
-def _make_read_key(
-    base_vault_url: str,
-    headers: Callable[[], Coroutine[Any, Any, dict]],
-):
-    async def read_key(path: str) -> Dict[str, str]:
-        return gamla.pipe(
-            await gamla.get_async_with_headers(
-                (await headers()),
-                30,
-                f"{base_vault_url}/secret/data/{path}",
-            ),
-            lambda response: response.json(),
-            gamla.excepts(
-                KeyError,
-                gamla.make_raise(KeyDoesNotExist),
-                gamla.get_in(["data", "data"]),
-            ),
+def _build_read_all_secrets(client: hvac.Client):
+    def read_all_secrets(path: str, version: Optional[str]) -> dict[str, str]:
+        read_secrets = client.secrets.kv.v2.read_secret_version(
+            path=path, version=version
         )
+        return read_secrets["data"]["data"]
 
-    return read_key
+    return read_all_secrets
 
 
-def _make_write_key(
-    base_vault_url: str,
-    headers: Callable[[], Coroutine[Any, Any, dict]],
-):
-    async def write_key(path: str, value: Dict[str, str]):
-        return await gamla.post_json_with_extra_headers_async(
-            (await headers()),
-            30,
-            f"{base_vault_url}/secret/data/{path}",
-            {"data": value},
+def _build_write_or_update_secrets(client: hvac.Client, read_all_secrets: Callable):
+    def write_or_update_secrets(path: str, secrets: dict[str, str]):
+        secrets_data = read_all_secrets(path)
+        secrets_data.update(secrets)
+        client.secrets.kv.v2.create_or_update_secret(
+            path=path,
+            secret=secrets_data,
+            mount_point=_MOUNT_POINT,
         )
 
-    return write_key
-
-
-def _make_update_key(read_key: Callable, write_key: Callable):
-    async def update_key(path: str, value: Dict[str, str]):
-        current_value = {}
-        try:
-            current_value = await read_key(path)
-        except KeyDoesNotExist:
-            pass
-        return await write_key(path, gamla.merge(current_value, value))
-
-    return update_key
-
-
-def make_vault(host: str, role: str, token: Optional[str]):
-    async def get_token():
-        return token
+    return write_or_update_secrets
 
-    base_vault_url = f"{host}/v1"
-    headers = _make_vault_headers(
-        get_token if token else _make_pod_identity_token(role, base_vault_url),
-    )
 
-    write_key, read_key = _make_write_key(base_vault_url, headers), _make_read_key(
-        base_vault_url,
-        headers,
-    )
-    return write_key, read_key, _make_update_key(read_key, write_key)
+def make_vault(host: str, token: Optional[str]):
+    client = hvac.Client(url=host, token=token)
```

Comment:
iiuc here we should have the Kubernetes authentication like in https://gitlab.com/airbud/devops/-/blob/main/vault_sync/sync_secondary_vault.py?ref_type=heads 

@JonSc8 is this in the scope of this task?

Reply:
@adi-rose yes it is - as discussed with @EdenSonbolyan, this will be the first step of the PR (just to review everything in the PR) and then we will add Kubernetes